### PR TITLE
New advanced ItemsRepeater Samples

### PR DESCRIPTION
**Pull request by anawishnoff**
## Description
Added two new ItemsRepeater samples to show off the more complex and unique features of the control. The first sample is a simple list that goes through a list of colors, but it uses custom animations to enlarge items as they get closer to the center. The second new sample is a custom, virtualizing layout for items of various sizes that are content-heavy. It shows how to sort, filter, and animate these items.

## Motivation and Context

Currently trying to increase ItemsRepeater's reach and share all of its advanced and unique functionality with our devs and community.

## How Has This Been Tested?

Tested on my own machine in light and dark mode, as well as with AccessibilityInsights for accessibility testing.

## Screenshots (if appropriate):
IRSample1
IRSample2

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
**Important:** In creating these samples some small bugs with ItemsRepeater and the custom layout functionality were found, causing the need to push a fix to the WinUI NuGet package. The code in this PR is running on the latest prerelease of the Microsoft.Ui.Xaml NuGet package (2.4.0-prerelease.200203002), which explains those changes to the .csproj file. See more at #342.

### CODE REVIEW

File: XamlControlsGallery/XamlControlsGallery.csproj [hunk 1]
```diff
@@ -416,9 +422,6 @@
     <Compile Include="ControlPages\InkCanvasPage.xaml.cs">
       <DependentUpon>InkCanvasPage.xaml</DependentUpon>
     </Compile>
-    <Compile Include="ControlPages\InkToolbarPage.xaml.cs">
```
Comment (stmoy) [inline XamlControlsGallery/XamlControlsGallery.csproj]: ? Please revert this change.

File: XamlControlsGallery/XamlControlsGallery.csproj [hunk 2]
```diff
@@ -11,7 +11,7 @@
     <DefaultLanguage>en-US</DefaultLanguage>
     <TargetPlatformIdentifier>UAP</TargetPlatformIdentifier>
     <TargetPlatformVersion>10.0.18362.0</TargetPlatformVersion>
-    <TargetPlatformMinVersion>10.0.17134.0</TargetPlatformMinVersion>
+    <TargetPlatformMinVersion>10.0.17763.0</TargetPlatformMinVersion>
```
Comment (stmoy) [inline XamlControlsGallery/XamlControlsGallery.csproj]: I understand needing to update the NuGet package. I don't understand why the minimum platform version is changed.

Comment (anawishnoff) [reply XamlControlsGallery/XamlControlsGallery.csproj]: About this and your other edit about the ink toolbar page, honestly not sure how these got added in - i may have pulled from an older version of the .csproj file. I'll fix these.

Comment (anawishnoff) [reply XamlControlsGallery/XamlControlsGallery.csproj]: Just to make sure before i change it back - @ranjeshj does the newest pre-release package require a higher/newer min platform version?

Comment (ranjeshj) [reply XamlControlsGallery/XamlControlsGallery.csproj]: Nope. No changes required for the platform versions.

File: XamlControlsGallery/Common/VariedImageSizeLayout.cs [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+using System.Collections.Generic;
+using System.Collections.Specialized;
+using System.Diagnostics;
+using Windows.Foundation;
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Controls;
+
+using VirtualizingLayout = Microsoft.UI.Xaml.Controls.VirtualizingLayout;
+using VirtualizingLayoutContext = Microsoft.UI.Xaml.Controls.VirtualizingLayoutContext;
+
```
Comment (ranjeshj) [inline XamlControlsGallery/Common/VariedImageSizeLayout.cs]: I think you can safely remove these.

File: XamlControlsGallery/Common/VariedImageSizeLayout.cs [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+namespace AppUIBasics.Common
+{
+    public class VariedImageSizeLayout : VirtualizingLayout
+    {
+        public VariedImageSizeLayout()
+        {
+            Width = 150.0;
+        }
+
+        public double Width { get; set; }
```
Comment (ranjeshj) [inline XamlControlsGallery/Common/VariedImageSizeLayout.cs]: you might be able to intialize to 150 here by saying = 150; and remove the one in the ctor.

File: XamlControlsGallery/Common/VariedImageSizeLayout.cs [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+            m_cachedBounds.Clear();
+            m_firstIndex = m_lastIndex = 0;
+            cachedBoundsInvalid = true;
+            InvalidateMeasure();
+        }
+
+        protected override Size MeasureOverride(VirtualizingLayoutContext context, Size availableSize)
+        {
+            var viewport = context.RealizationRect;
+            //Debug.WriteLine("Measure: " + viewport);
```
Comment (ranjeshj) [inline XamlControlsGallery/Common/VariedImageSizeLayout.cs]: remove or put behind an ifdef maybe ?

File: XamlControlsGallery/Common/VariedImageSizeLayout.cs [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+            if (availableSize.Width != m_lastAvailableWidth || cachedBoundsInvalid)
+            {
+                UpdateCachedBounds(availableSize);
+                m_lastAvailableWidth = availableSize.Width;
+                cachedBoundsInvalid = false;
```
Comment (ranjeshj) [inline XamlControlsGallery/Common/VariedImageSizeLayout.cs]: this can go into the UpdateCachedBounds method.

File: XamlControlsGallery/Common/VariedImageSizeLayout.cs [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+    public class VariedImageSizeLayout : VirtualizingLayout
+    {
+        public VariedImageSizeLayout()
+        {
+            Width = 150.0;
+        }
+
+        public double Width { get; set; }
+
+        private bool cachedBoundsInvalid = false;
```
Comment (ranjeshj) [inline XamlControlsGallery/Common/VariedImageSizeLayout.cs]: move these along with the other private fields below.

File: XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml [hunk 8]
```diff
@@ -74,6 +74,22 @@
         <common:ActivityFeedLayout x:Key="MyFeedLayout" ColumnSpacing="12" RowSpacing="12" MinItemSize="80, 108"/>
         <muxc:UniformGridLayout x:Key="UniformGridLayout2" MinItemWidth="108" MinItemHeight="108" MinRowSpacing="12" MinColumnSpacing="12"/>
 
+        <DataTemplate x:DataType="l:Recipe" x:Key="RecipeTemplate">
+            <StackPanel BorderThickness="1" Background="LightGray" Margin="5">
```
Comment (ranjeshj) [inline XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml]: switch to one of the existing theme resource. This likely wont look correct in one of the themes.

File: XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml [hunk 9]
```diff
@@ -188,7 +204,7 @@ the following attributes: Height, MaxHeight, Length, MaxLength, Diameter, and Ma
                 <local:ControlExampleSubstitution x:Name="SampleCodeLayout2" Key="SampleCodeLayout2"/>
             </local:ControlExample.Substitutions>
         </local:ControlExample>
-
```
Comment (ranjeshj) [inline XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml]: nit: spare newlines.

File: XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml [hunk 10]
```diff
@@ -212,14 +228,82 @@ the following attributes: Height, MaxHeight, Length, MaxLength, Diameter, and Ma
...
+                    <ColumnDefinition Width="1*"/>
+                </Grid.ColumnDefinitions>
+
+                <ScrollViewer x:Name="Animated_ScrollViewer" 
+                              Grid.Column="0"
+                              Height="175"
+                              Width="250"
+                              ViewChanging="Animated_ScrollViewer_ViewChanging">
+                    <muxc:ItemsRepeater x:Name="animatedScrollRepeater">
+                        <muxc:ItemsRepeater.ItemTemplate>
```
Comment (ranjeshj) [inline XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml]: nit: This line is technically not needed because it is the content property for ItemsRepeater.

File: XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml.cs [hunk 11]
```diff
@@ -52,49 +56,49 @@ private void InitializeData()
...
-                new NestedCategory("Proteins", new ObservableCollection<string>{
-                                                            "Steak",
-                                                            "Chicken",
-                                                            "Tofu",
-                                                            "Salmon",
-                                                            "Pork",
-                                                            "Chickpeas",
-                                                            "Eggs"
-                }));
+            ObservableCollection<string> fruits = new ObservableCollection<string>{
```
Comment (ranjeshj) [inline XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml.cs]: you can use initialization syntax to make this look more like the nested collection structure in code instead of adding one by one.

File: XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml.cs [hunk 12]
```diff
@@ -110,8 +114,67 @@ private void InitializeData()
...
+            colors.Add("DodgerBlue");
+            colors.Add("Firebrick");
+            colors.Add("Goldenrod");
+            colors.Add("Orange");
+            colors.Add("PaleVioletRed");
+            colors.Add("Olive");
+            colors.Add("MediumOrchid");
+            colors.Add("SeaGreen");
+            colors.Add("SandyBrown");
+            colors.Add("SteelBlue");
```
Comment (ranjeshj) [inline XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml.cs]: var colors = Enum.GetValues(typeof(Colors)).Cast<Color>().Select(c => c.ToString());

File: XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml.cs [hunk 13]
```diff
@@ -110,8 +114,67 @@ private void InitializeData()
...
+                rec.IngList = new List<string>() { fruitOption, vegOption, grainOption, proteinOption };
+
+                // Add extra ingredients so items have varied heights in the layout
+                rec.RandomizeIngredients();
+            }
+
+            // Initialize custom MyItemsSource object with new recipe data
+            filteredRecipeData.InitializeCollection(tempList);
+            // Save a static copy to compare to while filtering
+            staticRecipeData = new List<Recipe>(tempList);
```
Comment (ranjeshj) [inline XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml.cs]: can we just store tempList instead of creating a new List ?

File: XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml.cs [hunk 14]
```diff
@@ -313,4 +477,171 @@ public Bar(double length, int max)
...
     }
+
+    public class Recipe
+    {
+        public int Num { get; set; }
+        public string Ingredients { get; set; }
+        public List<string> IngList { get; set; }
+        public string Name { get; set; }
+        public string Color { get; set; }
+        public int numIngredients {
```
Comment (ranjeshj) [inline XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml.cs]: >numIngredients [](start = 19, length = 14)

capitalize N. Is this needed since it is just count of IngList ?

File: XamlControlsGallery/XamlControlsGallery.csproj (hunk 2 above)
Comment (anawishnoff) [reply XamlControlsGallery/XamlControlsGallery.csproj]: Thanks, changing back in upcoming commit.

File: XamlControlsGallery/XamlControlsGallery.csproj (hunk 1 above)
Comment (anawishnoff) [reply XamlControlsGallery/XamlControlsGallery.csproj]: Addded back in upcoming commit. I probably accidentally unloaded from project when i was trying to do so for the ItemsRepeater page.

File: XamlControlsGallery/Common/VariedImageSizeLayout.cs (hunk 3 above)
Comment (anawishnoff) [reply XamlControlsGallery/Common/VariedImageSizeLayout.cs]: It throws a bunch of errors when I remove these definitions (because I'm not using the Microsoft.Ui.Xaml in this file)?

File: XamlControlsGallery/Common/VariedImageSizeLayout.cs (hunk 6 above)
Comment (anawishnoff) [reply XamlControlsGallery/Common/VariedImageSizeLayout.cs]: @ranjeshj  Which code exactly? Lines 44-45 and if the if statement?

File: XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml (hunk 10 above)
Comment (anawishnoff) [reply XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml]: Cool, didn't know that! Fixed in upcoming commit.

File: XamlControlsGallery/Common/VariedImageSizeLayout.cs (hunk 6 above)
Comment (ranjeshj) [reply XamlControlsGallery/Common/VariedImageSizeLayout.cs]: oh i meant the cachedBoundsInvalid = false can move into the UPdateCacheBounds method, since calling that method essentially makes it valid.

File: XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml.cs (hunk 11 above)
Comment (anawishnoff) [reply XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml.cs]: What do you mean exactly by adding one by one? I thought I was using the object initialization syntax with collection<T> foo = new collection<T> { item array }

File: XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml.cs (hunk 12 above)
Comment (anawishnoff) [reply XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml.cs]: This exactly didnt work, but I changed it in an upcoming commit to getting all of the colors into a list in a similar way.

File: XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml.cs (hunk 14 above)
Comment (anawishnoff) [reply XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml.cs]: removed this

**Comment by anawishnoff** [comment]
@stmoy @ranjeshj Just wanted to bring this back up to see if it's ready to be merged!

File: XamlControlsGallery/XamlControlsGallery.csproj [hunk 15]
```diff
@@ -884,6 +890,10 @@
       <SubType>Designer</SubType>
       <Generator>MSBuild:Compile</Generator>
     </Page>
+    <Page Include="ControlPages\InkToolbarPage.xaml">
```
Comment (stmoy) [inline XamlControlsGallery/XamlControlsGallery.csproj]: This change isn't expected, right?

Comment (anawishnoff) [reply XamlControlsGallery/XamlControlsGallery.csproj]: I referred to the .csproj file in the master branch and saw that this was included, so I thought I should keep it included here as well.

Comment (stmoy) [reply XamlControlsGallery/XamlControlsGallery.csproj]: The trouble with the change is that it looks like it's adding another entry into the Pages table (meaning we have both this entry AND the one on line [745](https://github.com/microsoft/Xaml-Controls-Gallery/blob/acfb732c96a18c30258d53bc7324af0b8fd4ad2c/XamlControlsGallery/XamlControlsGallery.csproj#L745)). My guess is that this won't build.

Comment (anawishnoff) [reply XamlControlsGallery/XamlControlsGallery.csproj]: Gosh sorry, not sure why all of this got so jumbled. You're right. I will fix that and push

Comment (stmoy) [reply XamlControlsGallery/XamlControlsGallery.csproj]: Thanks for cleaning this up!

**Comment by stmoy** [comment]
It looks like there is also a merge conflict with ItemsRepeaterPage.xaml.cs -- can you take a look?

File: XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml.cs [hunk 16]
```diff
@@ -1,24 +1,32 @@
...
-        private bool isHorizontal = false;
 
         public ObservableCollection<Bar> BarItems;
+        public MyItemsSource filteredRecipeData = new MyItemsSource(null);
+        public List<Recipe> staticRecipeData;
+        private bool IsSortDescending = false;
+        public List<Recipe> tempList;
 
+        private double AnimatedBtnHeight;
+        private Windows.UI.Xaml.Thickness AnimatedBtnMargin;
```
Comment (ranjeshj) [inline XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml.cs]: >Windows.UI.Xaml. [](start = 16, length = 16)

nit: not needed since there is already a using statement for Windows.UI.Xaml

File: XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml.cs (hunk 11 above)
Comment (ranjeshj) [reply XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml.cs]: You can take it to both levels here like so.
var nestedCategories = new List<NestedCategory>()
{
   new NestedCategory("Vegitables", new ObservableCollection<string>(){"Broccoli", "Spinach"... },
   new NestedCategory("Fruits", new ObservableCollection<string>(){"Banana", "Grapes"... },

}

File: XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml.cs [hunk 17]
```diff
@@ -107,8 +109,56 @@ private void InitializeData()
...
+
+            // Initialize list of colors for animated scrolling sample
+            IList<String> colors = (typeof(Colors).GetRuntimeProperties().Select(c => c.ToString())).ToList();
+            for (int i =0; i < colors.Count(); i++)
+            {
+                colors[i] = colors[i].Substring(17);
+
+            }
+            
+            animatedScrollRepeater.ItemsSource = colors;
```
Comment (ranjeshj) [inline XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml.cs]: >colors [](start = 49, length = 6)

creating the list of colors can be extracted into its own method to avoid seeing its details here..
so something like
animatedScrollRepeater.ItemsSource = GetListOfColors();

File: XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml.cs [hunk 18]
```diff
@@ -230,6 +235,152 @@ private void RadioBtn_Click(object sender, RoutedEventArgs e)
...
+            scaleExpresion.Expression = "1 - abs((svVisual.Size.Y/2 - scrollProperties.Translation.Y) - (item.Offset.Y + item.Size.Y/2))*(.25/(svVisual.Size.Y/2))";
+
+            // Animate the item to change size based on distance from center of viewpoint
+            item.StartAnimation("Scale.X", scaleExpresion);
+            item.StartAnimation("Scale.Y", scaleExpresion);
+            var centerPointExpression = scrollProperties.Compositor.CreateExpressionAnimation();
+            centerPointExpression.SetReferenceParameter("item", item);
+            centerPointExpression.Expression = "Vector3(item.Size.X/2, item.Size.Y/2, 0)";
+            item.StartAnimation("CenterPoint", centerPointExpression);
+        }
```
Comment (ranjeshj) [inline XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml.cs]: nit: add newline

File: XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml.cs (hunk 17 above)
Comment (anawishnoff) [reply XamlControlsGallery/ControlPages/ItemsRepeaterPage.xaml.cs]: Cleaned it up, but put the recipe generator method outside the Recipe class as it needed to use the colors as well

File: XamlControlsGallery/XamlControlsGallery.csproj [hunk 19]
```diff
@@ -1141,7 +1147,7 @@
       <Version>6.1.9</Version>
     </PackageReference>
     <PackageReference Include="Microsoft.UI.Xaml">
-      <Version>2.4.0-prerelease.200203002</Version>
+      <Version>2.4.0-pr.200128002</Version>
```
Comment (stmoy) [inline XamlControlsGallery/XamlControlsGallery.csproj]: Does this update the nuget package to a lower version? Specifically to 20-01-28 (from 20-02-03)? Is this intentional?

Comment (anawishnoff) [reply XamlControlsGallery/XamlControlsGallery.csproj]: Just pushed an update with the latest pre-release nuget package. I was using the old one as it had the patch that Ranjesh pushed but it should be wrapped into the newest updated package by now (also tested it with the new package and it works).
